### PR TITLE
State fix2

### PR DESCRIPTION
**Pull request by machow**
Passed tests on validator, need to finish cleaning up code (see issue #101).

### CODE REVIEW

**Comment by filipsch** [comment]
@machow i'll wait with reviewing until you did the clean up, ping me when done!

**Comment by filipsch** [comment]
@machow When I run the tests, there's still a large amount of exercises that fail. Some excerpts:

```
======================================================================
ERROR: test_Pass (test_test_intermediate.TestExercise6)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/filip/workspace/pythonwhat/tests/test_test_intermediate.py", line 443, in test_Pass
    sct_payload = helper.run(self.data)
  File "/Users/filip/workspace/pythonwhat/tests/helper.py", line 21, in run
    output = exercise.runSubmit(data)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/pythonbackend/Exercise.py", line 124, in runSubmit
    raw_output["error"]
  File "/Users/filip/workspace/pythonwhat/pythonwhat/test_exercise.py", line 68, in test_exercise
    test(state)
  File "/Users/filip/workspace/pythonwhat/pythonwhat/probe.py", line 102, in __call__
    self.data['func'](state=state, *ba.args, **ba.kwargs)
  File "/Users/filip/workspace/pythonwhat/pythonwhat/test_funcs/test_expression_result.py", line 92, in test_expression_result
    raise ValueError("Running the expression in the solution process caused an error.")
ValueError: Running the expression in the solution process caused an error.
```

```
======================================================================
ERROR: test_pass (test_test_if_else.TestIfExpListComp)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/filip/workspace/pythonwhat/tests/test_test_if_else.py", line 393, in test_pass
    sct_payload = helper.run(self.data)
  File "/Users/filip/workspace/pythonwhat/tests/helper.py", line 21, in run
    output = exercise.runSubmit(data)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/pythonbackend/Exercise.py", line 124, in runSubmit
    raw_output["error"]
  File "/Users/filip/workspace/pythonwhat/pythonwhat/test_exercise.py", line 68, in test_exercise
    test(state)
  File "/Users/filip/workspace/pythonwhat/pythonwhat/probe.py", line 102, in __call__
    self.data['func'](state=state, *ba.args, **ba.kwargs)
  File "/Users/filip/workspace/pythonwhat/pythonwhat/test_funcs/test_expression_result.py", line 94, in test_expression_result
    raise ValueError("The result of running the expression in the solution process couldn't be figured out: " + eval_solution.info)
ValueError: The result of running the expression in the solution process couldn't be figured out: dilling inside process failed for None - write manual converter
```

I made some small changes here and there and did some cleanup, but this cannot be the reason of these errors. 
What am I doing wrong? For you all the tests run without issues?

**Comment by filipsch** [comment]
The tests pass on Jenkins though, so not sure what's going on.

**Comment by machow** [comment]
I just pulled your commits and was able to run the tests. Did you try cloning a fresh copy of the repo? Things I could think of causing the difference on your machine (in order of ease of checking?!)...

1. Different version of the python backend than Jenkins
2. Different python version (I'm using 3.5.2, but 3.6 recently came out)
3. Different library versions (it looks like pythonwhat doesn't specify what versions of ipython it's compatible with, for instance).

We can check it out whenever we meet today!

File: pythonwhat/probe.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+            nl = NodeList(name = "ListDeferred", arg_name = arg_name)
+            node.add_child(nl)
+            prev_node, tree.crnt_node = tree.crnt_node, nl
+            test()
+            tree.crnt_node = prev_node
+        elif test is not None:
+            raise Exception("Expected a function or list/tuple/dict of functions")
+
+
+def create_test_probes(test_exercise):
```
Comment (filipsch) [inline pythonwhat/probe.py]: I find `test_exercise` a confusing argument here. Can you use a name that refers to the fact it is a context?

File: pythonwhat/probe.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    "test_object_accessed",
+    "test_output_contains",
+    "test_lambda_function",
+    "test_expression_result",
+    "test_expression_output",
+    "test_function_definition",
+    "test_object_after_expression"
+]
+
+SUB_TESTS = {
```
Comment (filipsch) [inline pythonwhat/probe.py]: I see that in line 157, you put as TODO that you want to make this automatic. How hard is this?

File: pythonwhat/probe.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        self.tree.crnt_node.add_child(this_node)
+
+        da = bound_args.arguments
+        for st in self.sub_tests:     # TODO: auto sub test detection
+            if st in da and da[st]:
+                self.run_sub_tests(da[st], self.tree, this_node, st)
+        return this_node
+
+    @staticmethod
+    def run_sub_tests(test, tree, node, arg_name):
```
Comment (filipsch) [inline pythonwhat/probe.py]: Is `run_sub_tests()` a good name? It's not actually running tests, is it? It's _building_ nodes?

File: pythonwhat/probe.py (hunk 2 above)
Comment (machow) [inline pythonwhat/probe.py]: An easy way would be to assume that functions (or functions w/o arguments) and Nodes are sub_tests, but there are also cases where dictionaries and lists of functions are valid sub-tests.

Alternatively, I thought about modifying the test_* functions like...

```{python}
from sub_test import SubTestMarker

def with(STUFF..., body=SubTestMarker, state=None):
    # CODE...
```

Where SubTestMarker marks that `body` is a sub-test argument. Currently the default arguments for subtests are None, so this is like creating a sort of NoneSubtest type.

**Pull request merged**